// astro/dist/vite-plugin-config-alias/index.js
import fs from "node:fs";
import path from "node:path";
import { normalizePath } from "vite";
const getConfigAlias = (settings) => {
  const { tsConfig, tsConfigPath } = settings;
  if (!tsConfig || !tsConfigPath || !tsConfig.compilerOptions) return null;
  const { baseUrl, paths } = tsConfig.compilerOptions;
  const effectiveBaseUrl = baseUrl ?? (paths ? "." : void 0);
  if (!effectiveBaseUrl) return null;
  const resolvedBaseUrl = path.resolve(path.dirname(tsConfigPath), effectiveBaseUrl);
  const aliases = [];
  if (paths) {
    for (const [alias, values] of Object.entries(paths)) {
      const find = new RegExp(
        `^${[...alias].map(
          (segment) => segment === "*" ? "(.+)" : segment.replace(/[\\^$*+?.()|[\]{}]/, "\\$&")
        ).join("")}$`
      );
      for (const value of values) {
        let matchId = 0;
        const replacement = [...normalizePath(path.resolve(resolvedBaseUrl, value))].map((segment) => segment === "*" ? `$${++matchId}` : segment === "$" ? "$$" : segment).join("");
        aliases.push({ find, replacement });
      }
    }
  }
  if (baseUrl) {
    aliases.push({
      find: /^(?!\.*\/|\.*$|\w:)(.+)$/,
      replacement: `${[...normalizePath(resolvedBaseUrl)].map((segment) => segment === "$" ? "$$" : segment).join("")}/$1`
    });
  }
  return aliases;
};
function resolveWithAlias(id, configAlias) {
  for (const alias of configAlias) {
    if (alias.find.test(id)) {
      const updatedId = id.replace(alias.find, alias.replacement);
      const stats = fs.statSync(updatedId, { throwIfNoEntry: false });
      if (stats && stats.isFile()) {
        return normalizePath(updatedId);
      }
    }
  }
  return null;
}
const cssImportRE = /@import\s+(?:url\(\s*)?['"]([^'"]+)['"]\s*\)?/g;
function configAliasVitePlugin({
  settings
}) {
  const configAlias = getConfigAlias(settings);
  if (!configAlias) return null;
  return [
    // Pre-plugin: rewrite CSS @import aliases to absolute paths before Vite's CSS plugin.
    // Vite's internal CSS @import resolver (postcss-import) uses a mini plugin container
    // that doesn't include user resolveId hooks, so we must rewrite aliases in a transform
    // hook that runs before Vite's CSS processing.
    {
      name: "astro:tsconfig-alias-css",
      enforce: "pre",
      transform: {
        filter: {
          id: {
            include: /\.css$/
          }
        },
        handler(code) {
          if (!code.includes("@import")) return;
          let hasReplacement = false;
          const result = code.replace(cssImportRE, (match, importId) => {
            if (!importId) return match;
            const resolved = resolveWithAlias(importId, configAlias);
            if (resolved) {
              hasReplacement = true;
              return match.replace(importId, resolved);
            }
            return match;
          });
          if (hasReplacement) {
            return { code: result, map: null };
          }
        }
      }
    },
    // Post-plugin: resolve JS/TS imports using tsconfig path aliases via resolveId.
    {
      name: "astro:tsconfig-alias",
      // use post to only resolve ids that all other plugins before it can't
      enforce: "post",
      resolveId: {
        filter: {
          id: {
            include: configAlias.map((alias) => alias.find),
            exclude: /(?:\0|^virtual:|^astro:)/
          }
        },
        async handler(id, importer, options) {
          for (const alias of configAlias) {
            if (alias.find.test(id)) {
              const updatedId = id.replace(alias.find, alias.replacement);
              if (updatedId.includes("*")) {
                return updatedId;
              }
              const resolved = await this.resolve(updatedId, importer, {
                skipSelf: true,
                ...options
              });
              if (resolved) return resolved;
            }
          }
        }
      }
    }
  ];
}
export {
  configAliasVitePlugin as default
};
